src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -31,109 +31,153 @@
 
+            pets.add(pet);
+
+        } while (currentPosition < csv.length() && pet != null);
+
+        return new ResponseEntity<List<Pet>>(pets, HttpStatus.OK);
+    }
+
+    public class Result {
+        public Result(int currPosition) {
             field = "";
-            while (i < csv.length() && (csv.charAt(i) != ';' && csv.charAt(i) != '\n')) {
-                field += csv.charAt(i++);
-            }
+            position = currPosition;
+        }
+        public String field;
+        public int position;
+    }
 
-            if (pet != null) {
-                String owner = field;
-                List<Owner> matchingOwners = clinicService.findAllOwners()
-                    .stream()
-                    .filter(o -> o.getLastName().equals(owner))
-                    .collect(Collectors.toList());
-
-                if (matchingOwners.size() == 0) {
-                    HttpHeaders headers = new HttpHeaders();
-                    headers.add("errors", "Owner not found");
-                    return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
-                }
-                if (matchingOwners.size() > 1) {
-                    HttpHeaders headers = new HttpHeaders();
-                    headers.add("errors", "Owner not unique");
-                    return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
+    private Result getNextWord(String csv, int currPosition, Character[] delimiters) {
+        Result result = new Result(currPosition);
+        if (result.position < csv.length()) {
+            boolean canAdvance = true;
+            for (Character delimiter: delimiters) {
+                if (csv.charAt(result.position) == delimiter) {
+                    canAdvance = false;
+                    break;
                 }
-                pet.setOwner(matchingOwners.iterator().next());
             }
+            if (canAdvance) {
+                result.field += csv.charAt(result.position++);
+            }
+        }
+        return result;
+    }
 
-            if (csv.charAt(i) == ';') {
-                i++;
+    private int extractAndSetName(String csv, int currPosition, Pet pet) {
+        Character[] semicolon = {';'};
+        Result result = getNextWord(csv, currPosition, semicolon);
+        pet.setName(result.field);
+        return ++result.position;
+    }
 
-                field = "";
-                while (i < csv.length() && csv.charAt(i) != '\n') {
-                    field += csv.charAt(i++);
-                }
+    private int extractAndSetBirthDate(String csv, int currPosition, Pet pet) throws ParseException {
+        Character[] semicolon = {';'};
+        Result result = getNextWord(csv, currPosition, semicolon);
+        try {
+            pet.setBirthDate((new SimpleDateFormat("yyyy-MM-dd")).parse(result.field));
+        } catch (ParseException e) {
+            throw new ParseException("date " + result.field + " not valid", result.position);
```

Comment:
ah...you actually propagating the error here upwards, which is nice :) then in the above `importPets` method, you can ease the error handling now